### PR TITLE
Portrait Mode BMD

### PR DESCRIPTION
**Pull request by beausmith**
## Overview
Resolves [#2522](https://github.com/votingworks/vxsuite/issues/2522)

Related: https://github.com/votingworks/roadmap/issues/40

## Demo Video or Screenshot
https://user-images.githubusercontent.com/28444/207462356-b56e1b04-2042-4e07-a019-643517b2e3c3.mov

## Checklist
- [x] I have added a screenshot and/or video to this PR to demo the change
- [x] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

File: frontends/bmd/src/pages/contest_page.tsx [hunk 1]
```diff
@@ -107,67 +171,68 @@ export function ContestPage(): JSX.Element {
...
+              primary
+              onPress={() => {}}
+              aria-label="Change Language"
+            >
+              Español
+            </Button>
+            <Button large onPress={() => {}} aria-label="Change Language">
+              English
+            </Button>
+          </SegmentedButton> */}
```
Comment (adghayes) [inline frontends/bmd/src/pages/contest_page.tsx:193]: Should this comment be removed?

File: frontends/bmd/src/pages/review_page.tsx [hunk 2]
```diff
@@ -516,38 +528,46 @@ export function ReviewPage(): JSX.Element {
...
-            </LinkButton>
-          </p>
-        </Prose>
-      </Sidebar>
+      {isPortrait ? (
+        <ButtonFooter>
+          {printMyBallotButton}
+          {/* <Button large onPress={() => {}}>
+            English
+          </Button> */}
```
Comment (adghayes) [inline frontends/bmd/src/pages/review_page.tsx:536]: Should you remove this comment? Or are you leaving the "English" ones in for later work?

File: frontends/bmd/src/pages/review_page.tsx [hunk 3]
```diff
@@ -516,38 +528,46 @@ export function ReviewPage(): JSX.Element {
...
+        </ButtonFooter>
+      ) : (
+        <Sidebar
+          footer={
+            <React.Fragment>
+              <SettingsTextSize
+                userSettings={userSettings}
+                setUserSettings={setUserSettings}
+                sidebarWrapper
+              />
```
Comment (adghayes) [inline frontends/bmd/src/pages/review_page.tsx:553]: We only have one setting right now, but as we add other settings, I'm thinking we'll want some sort of wrapper component containing all the settings to put in the sidebar in landscape mode. So we don't have different ordering or forget to add settings to a page. Thoughts?

Comment (adghayes) [reply frontends/bmd/src/pages/review_page.tsx:553]: Not suggesting a change for this PR.

File: frontends/bmd/src/pages/user_settings_modal.tsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      }
+      if (event.key === 'ArrowLeft') {
+        event.stopPropagation();
+        handleGamepadKeyboardEvent({ ...event, key: 'ArrowUp' });
+      }
+    }
+    document.addEventListener('keydown', handleKeyDown, { capture: true });
+    return () =>
+      document.removeEventListener('keydown', handleKeyDown, { capture: true });
+  }, []);
```
Comment (adghayes) [inline frontends/bmd/src/pages/user_settings_modal.tsx:34]: Could you ELI5 why this is necessary?

File: libs/ui/src/election_info_bar.tsx [hunk 5]
```diff
@@ -37,49 +38,45 @@ export function ElectionInfoBar({
   machineId,
   precinctSelection,
 }: Props): JSX.Element {
-  const electionDate = format.localeDate(
-    new Date(electionDefinition.election.date)
-  );
+  const {
+    election: { precincts, date, title, county, state, seal, sealUrl },
```
Comment (adghayes) [inline libs/ui/src/election_info_bar.tsx:42]: Woah, didn't know you could do nested destructuring...

File: libs/ui/src/election_info_bar.tsx [hunk 6]
```diff
@@ -37,49 +38,45 @@ export function ElectionInfoBar({
...
+  } = electionDefinition;
+  const electionDate = format.localeDate(new Date(date));
 
   /* istanbul ignore next */
   const theme: Theme = { ...(contrastTheme.dark ?? {}) };
 
   return (
     <Bar theme={theme} data-testid="electionInfoBar">
+      {(seal || sealUrl) && <Seal seal={seal} sealUrl={sealUrl} />}
```
Comment (adghayes) [inline libs/ui/src/election_info_bar.tsx:51]: Not in this PR, but I think we should just do away with non `sealUrl` seals because of the added complexity.

File: libs/ui/src/modal.tsx [hunk 7]
```diff
@@ -123,6 +124,17 @@ function focusModalAudio() {
   }, 10);
 }
 
+/* istanbul ignore next - unclear why this isn't covered */
```
Comment (adghayes) [inline libs/ui/src/modal.tsx:127]: Perhaps the test doesn't wait for the timeout to run?

**Review by adghayes — COMMENTED** [review]
@beausmith I'm summarizing the high-level component changes here. Let me know if I'm misunderstanding. One outstanding suggestion with the `ReplaceElectionScreen`.

**Voting Screens**
Pattern: use `Sidebar` in landscape mode and `ButtonFooter` or `Footer` in portrait mode. This is to have ideal voting experience, esp. in terms of button location. The complexity of maintaining this seems the highest, but necessary. It does mean that now we have different information shown on screen in the landscape and portrait mode - e.g. `ElectionInfo` is shown in landscape mode and not in portrait mode.

**Non-Voting Screens**
Pattern: use `ElectionInfoBar` in both landscape and portrait mode. This is nice because it is less complex and we can use a shared `libs/ui` component. My only UI nit is that the blue of our footer looks weird to me in an app where the rest of our footers and sidebars are black or gray, but this will all have to change with contrast ratios anyway.

Would it make sense to use `ElectionInfoBar` on the `ReplaceElectionScreen`? Currently it's the only non-voting screen that's still using `Sidebar`. It would help simplify the pattern in my mind if `Sidebar` was exclusively a voting component.

**Diagnostic Screen**
You removed the sidebar because it's not necessary, makes sense.

File: frontends/bmd/src/components/settings_text_size.tsx [hunk 8]
```diff
@@ -39,49 +41,56 @@ const FontSizeButtons = styled.div`
...
   setUserSettings: SetUserSettings;
+  sidebarWrapper?: boolean;
 }
 
+const ariaLabels = ['Small', 'Medium', 'Large', 'Extra Large'];
+
 export function SettingsTextSize({
   userSettings,
   setUserSettings,
+  sidebarWrapper = false,
```
[the remark this answers is not in the document]
Comment (adghayes) [reply frontends/bmd/src/components/settings_text_size.tsx:60]: I understand now that this is indicates that the component should be wrapped in a container making it appropriate for being part of a sidebar. What do you think of creating a different component entirely, a sidebar settings wrapper, that would also address this comment? https://github.com/votingworks/vxsuite/pull/2855#discussion_r1048803385

File: frontends/bmd/src/app_end_to_end.test.tsx [hunk 9]
```diff
@@ -241,17 +236,43 @@ it('MarkAndPrint end-to-end flow', async () => {
   await advanceTimersAndPromises();
   expect(writeLongUint8ArrayMock).toHaveBeenCalledTimes(1);
 
+  // Adjust Text Size in Settings Modal
+  fireEvent.click(screen.getByText('Settings'));
+  screen.getByText('Voter Settings');
+  expect(screen.getAllByLabelText('Text Size:', { exact: false }).length).toBe(
+    3
+  );
+  act(() => void userEvent.keyboard('{ArrowRight}'));
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/app_end_to_end.test.tsx:245]: This was copy/pasta from another test. I will attempt to remove the act and void to see if I get any errors.

File: frontends/bmd/src/app_end_to_end.test.tsx [hunk 10]
```diff
@@ -241,17 +236,43 @@ it('MarkAndPrint end-to-end flow', async () => {
   await advanceTimersAndPromises();
   expect(writeLongUint8ArrayMock).toHaveBeenCalledTimes(1);
 
+  // Adjust Text Size in Settings Modal
+  fireEvent.click(screen.getByText('Settings'));
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/app_end_to_end.test.tsx:240]: Will update to recommended.

File: frontends/bmd/src/app_root.tsx [hunk 11]
```diff
@@ -133,17 +133,18 @@ export const stateStorageKey = 'state';
...
 };
 
 const initialHardwareState: Readonly<HardwareState> = {
   machineConfig: {
-    appMode: MarkOnly,
+    appMode: MarkAndPrint,
     machineId: '0000',
     codeVersion: 'dev',
     screenOrientation: 'portrait',
+    // screenOrientation: 'landscape',
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/app_root.tsx:147]: yes

File: frontends/bmd/src/app_root.tsx [hunk 12]
```diff
@@ -242,9 +243,9 @@ function appReducer(state: State, action: AppAction): State {
       };
     case 'setUserSettings':
       /* istanbul ignore next */
-      if (Object.keys(action.userSettings).join(',') !== 'textSize') {
-        throw new Error('unknown userSetting key');
-      }
+      // if (Object.keys(action.userSettings).join(',') !== 'textSize') {
+      //   throw new Error('unknown userSetting key');
+      // }
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/app_root.tsx:248]: yes

File: frontends/bmd/src/app_root.tsx [hunk 13]
```diff
@@ -431,8 +432,10 @@ export function AppRoot({
 
   const hidePostVotingInstructions = useCallback(() => {
     clearTimeout(PostVotingInstructionsTimeout.current);
+    // if (false) {
     if (isCardlessVoterAuth(auth)) auth.logOut();
     dispatchAppState({ type: 'resetBallot' });
+    // }
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/app_root.tsx:438]: yes

File: frontends/bmd/src/components/election_info.test.tsx [hunk 14]
```diff
@@ -46,3 +62,11 @@ it('renders vertical ElectionInfo without hash by default', () => {
...
   expect(container).toMatchSnapshot();
 });
+
+// test.only('renders vertical without precinct name', () => {
+//   render(<ElectionInfo electionDefinition={electionSampleDefinition} />);
+//   screen.debug();
+//   screen.getByLabelText(
+//     'November 3, 2020. State of Hamilton, Franklin County.'
+//   );
+// });
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/components/election_info.test.tsx:72]: yes

File: frontends/bmd/src/components/settings_text_size.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/components/settings_text_size.tsx:60]: Yes, could create a separate component… though I my thought was to make this easy to delete when we remove the landscape format in the (hopefully) near future.

File: frontends/bmd/src/components/yes_no_contest.test.tsx [hunk 15]
```diff
@@ -44,7 +47,8 @@ describe('supports yes/no contest', () => {
       <YesNoContest contest={contest} vote={['yes']} updateVote={updateVote} />
     );
     expect(container).toMatchSnapshot();
-    fireEvent.click(screen.getByText('No').closest('button')!);
+    const contestChoices = screen.queryByTestId('contest-choices');
+    fireEvent.click(within(contestChoices!).getByText('No').closest('button')!);
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/components/yes_no_contest.test.tsx:51]: 🤦  Thanks. Will do.

File: frontends/bmd/src/pages/contest_page.tsx (hunk 1 above)
Comment (beausmith) [reply frontends/bmd/src/pages/contest_page.tsx:193]: No, I had thought to leave this in as example of how to add language. Thoughts?

File: frontends/bmd/src/pages/review_page.tsx (hunk 2 above)
Comment (beausmith) [reply frontends/bmd/src/pages/review_page.tsx:536]: Same as above comment. (No, I had thought to leave this in as example of how to add language. Thoughts?) But perhaps I should create a shared component for these two comments… or just delete it all together?

File: frontends/bmd/src/pages/review_page.tsx (hunk 3 above)
Comment (beausmith) [reply frontends/bmd/src/pages/review_page.tsx:553]: Most-likely yes. However there are questions to answer here:
- How long will landscape orientation be supported?
- Do we want to modify the landscape orientation based upon what we've implemented for the portrait orientation?
- Will the landscape orientation use "Settings" and languages buttons similar to how the portrait orientation UI works?

File: frontends/bmd/src/pages/user_settings_modal.tsx (hunk 4 above)
Comment (beausmith) [reply frontends/bmd/src/pages/user_settings_modal.tsx:34]: In the Settings modal, pressing the right and left arrows navigates to the next/previous contests behind the settings modal. The settings modal UI has a bit of horizontal UI, so we could do one of two things:
- map the left/right to the up/down so that it works as a user might expect (as implemented)
- disable the left/right when the settings modal is option (might be a better solution)

File: libs/ui/src/election_info_bar.tsx (hunk 6 above)
Comment (beausmith) [reply libs/ui/src/election_info_bar.tsx:51]: I'm on the fence. 

## Inline Seal
- Full election definition portability
- Can use our products without first having to commit your seal to our repo.

## Seal URL
- There is simplicity with only referring to seals which are added to the codebase.
- Only seals in the codebase can be used.
- Assets have to be added to every app (currently, though perhaps a build process could solve this?)

File: frontends/bmd/src/app_end_to_end.test.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/app_end_to_end.test.tsx:245]: Works! I admit I'm still worried about those act warnings that the texts throw sometimes.

File: frontends/bmd/src/components/settings_text_size.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/components/settings_text_size.tsx:60]: For consistency, I'm going to replace the Text Size settings in the sidebar (horizontal orientation) with a link to the same Voter Settings modal.

**Review by mattroe — COMMENTED** [review]
Overall LGTM. A few minor details noticed when testing on a VxDev VxMark (Lenovo Flex) in a vertical orientation:

### Footer on Poll Worker Screens

Long election name makes the footer super tall. I would suggest this name takes up more horizontal real estate like the unauthenticated view.
Screenshot from 2022-12-14 19-21-41

### Accessibility Controller Test Screen
UI is pretty broken. Other diagnostic screens looked fine.
Screenshot from 2022-12-14 19-23-25

### Review Screen
Largest text size breaks the UI a bit. Other text sizes looked fine. OK to pick up in the follow up text size setting work if easier.
Screenshot from 2022-12-14 19-25-46

### Post-Activation Screen
Could use a bit more left side padding
Screenshot from 2022-12-14 19-28-50

File: frontends/bmd/src/pages/review_page.tsx [hunk 16]
```diff
@@ -546,11 +548,29 @@ export function ReviewPage(): JSX.Element {
...
+                  >
+                    Español
+                  </Button>
+                  <Button onPress={() => {}} aria-label="Change Language">
+                    English
+                  </Button>
+                </SegmentedButton> */}
+                {/* <Button onPress={() => {}} aria-label="Change Language">
+                  English
+                </Button> */}
```
Comment (adghayes) [inline frontends/bmd/src/pages/review_page.tsx:566]: Are these commented out buttons meant to be in a different order for landscape and portrait mode respectively?

File: frontends/bmd/src/pages/review_page.tsx [hunk 17]
```diff
@@ -546,11 +548,29 @@ export function ReviewPage(): JSX.Element {
...
+                <Button
+                  onPress={() => setUserSettings({ showSettingsModal: true })}
+                  aria-label="Change Settings"
+                >
+                  Settings
+                </Button>
```
Comment (adghayes) [inline frontends/bmd/src/pages/review_page.tsx:572]: Is it possible to pull out these settings as a `React.Fragment`?

Comment (adghayes) [reply frontends/bmd/src/pages/review_page.tsx:572]: In order to re-use in the landscape and portrait case.

**Review by adghayes — COMMENTED** [review]
Improvements look great. There are a couple places where I'd prefer to see components broken out.

The settings button exists in 4 places, and could be broken out into its own component: 
```jsx
<Button
  large
  onPress={() => setUserSettings({ showSettingsModal: true })}
  aria-label="Change Settings"
>
  Settings
</Button>
```
The related, commented out language options are a bit confusing because I see them in both orders (maybe that's due to landscape and portrait) and I see one place where there's only the non-segmented button. Should these all be the same? If so, could those comments be pulled out into a separate component which is a `React.Fragment` containing those comments and the settings button above?

On the `StartPage`, I think it would be beneficial to pull out some of the repeated components in portrait and landscape, e.g. the wobble start voting button and the settings container, as you have done with the `NextContestButton` on the contest page. 

My main concern is that, with support for both landscape and portrait (however long), I want to create fewer places where we could accidentally fix or change something in one mode and not the other. 

Finally, I haven't looked at the new Replace Election Screen. I'm not sure from just looking at the code if the contents of the screen have changed. Not a blocker, but if you have a screenshot that would be helpful.

File: frontends/bmd/src/pages/review_page.tsx (hunk 3 above)
Comment (adghayes) [reply frontends/bmd/src/pages/review_page.tsx]: I don't know how long landscape will be supported, and I would like to _not_ support it for long, but I have a suspicion is will stick around or come back up.

File: frontends/bmd/src/pages/user_settings_modal.tsx (hunk 4 above)
Comment (adghayes) [reply frontends/bmd/src/pages/user_settings_modal.tsx:34]: This seems plenty good, just wondering.

File: frontends/bmd/src/pages/review_page.tsx (hunk 3 above)
Comment (beausmith) [reply frontends/bmd/src/pages/review_page.tsx]: Agreed. FWIW, I tried to keep as little landscape-specific conditionals as possible. This one has been removed.

File: libs/ui/src/election_info_bar.tsx [hunk 18]
```diff
@@ -41,7 +42,10 @@ export function ElectionInfoBar({
   const {
     election: { precincts, date, title, county, state, seal, sealUrl },
   } = electionDefinition;
-  const electionDate = format.localeDate(new Date(date));
+  // const electionDate = format.localeDate(new Date(date));
+  //                  <td>{formatLongDate(DateTime.fromISO(election.date))}</td>
```
[the remark this answers is not in the document]
Comment (beausmith) [reply libs/ui/src/election_info_bar.tsx:46]: removed.

File: frontends/bmd/src/pages/start_page.tsx [hunk 19]
```diff
@@ -138,21 +144,14 @@ export function StartPage(): JSX.Element {
       ) : (
         <Sidebar
           footer={
-            <React.Fragment>
-              <SettingsTextSize
-                userSettings={userSettings}
-                setUserSettings={setUserSettings}
-                sidebarWrapper
+            isLandscape && (
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/pages/start_page.tsx:147]: yes 🤦

File: frontends/bmd/src/pages/start_page.tsx [hunk 20]
```diff
@@ -101,6 +101,12 @@ export function StartPage(): JSX.Element {
                 <strong>{pluralize('contest', contests.length, true)}</strong>.
               </span>
             </p>
+            <SettingsContainer>
+              <SettingsTextSize
+                userSettings={userSettings}
+                setUserSettings={setUserSettings}
+              />
+            </SettingsContainer>
```
[the remark this answers is not in the document]
Comment (beausmith) [reply frontends/bmd/src/pages/start_page.tsx:109]: Agreed. Done.

File: frontends/bmd/src/pages/review_page.tsx (hunk 17 above)
Comment (beausmith) [reply frontends/bmd/src/pages/review_page.tsx:572]: Agreed. Done.

File: frontends/bmd/src/pages/review_page.tsx (hunk 16 above)
Comment (beausmith) [reply frontends/bmd/src/pages/review_page.tsx:566]: Cleaned up by creating a component.

File: libs/types/src/generic.ts [hunk 21]
```diff
@@ -2,6 +2,9 @@ import check8601 from '@antongolub/iso8601';
 import { z } from 'zod';
 import { err, ok, Result, wrapException } from './result';
 
+export interface DefinedDictionary<T> {
+  [key: string]: T;
+}
```
Comment (adghayes) [inline libs/types/src/generic.ts:7]: I have been meaning to define a type like this, thanks!

**Review by adghayes — APPROVED** [review]
LGTM, thanks for all the work on this!

File: libs/types/src/generic.ts (hunk 21 above)
Comment (beausmith) [reply libs/types/src/generic.ts:7]: I would have changed the other to `OptionalDictionary` but 🤷‍♂️
